core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala

```diff
@@ -16,233 +16,1555 @@
+  /**
+   * Return the string which we want to print to describe the current partition assignment.
+   *
+   * @param proposedParts               The proposed partition assignment.
+   * @param currentParts                The current partition assignment.
+   *
+   * @return                            The string to print.  We will only print information about
+   *                                    partitions that appear in the proposed partition assignment.
+   */
+  def currentPartitionReplicaAssignmentToString(proposedParts: Map[TopicPartition, Seq[Int]],
+                                                currentParts: Map[TopicPartition, Seq[Int]]): String = {
+    "Current partition replica assignment%n%n%s%n%nSave this to use as the %s".
+        format(formatAsReassignmentJson(currentParts.filterKeys(proposedParts.contains(_)).toMap, Map.empty),
+              "--reassignment-json-file option during rollback")
+  }
+
+  /**
+   * Verify that the replicas and brokers referenced in the given partition assignment actually
+   * exist.  This is necessary when using the deprecated ZK API, since ZooKeeper itself can't
+   * validate what we're applying.
+   *
+   * @param zkClient                    The ZooKeeper client to use.
+   * @param proposedParts               The partition assignment.
+   */
+  def verifyReplicasAndBrokersInAssignment(zkClient: KafkaZkClient,
+                                           proposedParts: Map[TopicPartition, Seq[Int]]): Unit = {
+    // check that all partitions in the proposed assignment exist in the cluster
+    val proposedTopics = proposedParts.map { case (tp, _) => tp.topic }
+    val existingAssignment = zkClient.getReplicaAssignmentForTopics(proposedTopics.toSet)
+    val nonExistentPartitions = proposedParts.map { case (tp, _) => tp }.filterNot(existingAssignment.contains)
+    if (nonExistentPartitions.nonEmpty)
+      throw new AdminCommandFailedException("The proposed assignment contains non-existent partitions: " +
+        nonExistentPartitions)
+
+    // check that all brokers in the proposed assignment exist in the cluster
+    val existingBrokerIDs = zkClient.getSortedBrokerList
+    val nonExistingBrokerIDs = proposedParts.toMap.values.flatten.filterNot(existingBrokerIDs.contains).toSet
+    if (nonExistingBrokerIDs.nonEmpty)
+      throw new AdminCommandFailedException("The proposed assignment contains non-existent brokerIDs: " + nonExistingBrokerIDs.mkString(","))
+  }
+
+  /**
+   * Execute the given partition reassignments.
+   *
+   * @param adminClient       The admin client object to use.
+   * @param reassignments     A map from topic names to target replica assignments.
+   * @return                  A map from partition objects to error strings.
+   */
+  def alterPartitionReassignments(adminClient: Admin,
+                                  reassignments: Map[TopicPartition, Seq[Int]])
+                                
```

Comment:
I'm guessing a `None` is implied here? If so, could we include it?

Reply:
Github's broken UI is showing this comment as being on line 1170 for me.  Since the comment doesn't make sense for that line, what line did you intend this for?  I checked the file for `case Some` without `case None`, but didn't find anything...

Reply:
After your latest changes, this is L1421 inside `def modifyTopicThrottles`.

```scala
leaderThrottles.get(topicName) match {
  case None =>
  case Some(value) => ...
}
```

Reply:
OK, then it's already been addressed.  Sounds good.